### DISCUSSION TITLE
HttpResponse.redirect?

### DISCUSSION THREAD

**Discussion by dalisoft** [uNetworking/uWebSockets.js#156] [Q&A] [answered]
Hi @alexhultman 

Is there any way like `res.redirect(newPath)` or similar function? In examples folder i'm didn't see any example.

If there no method, any way realize this?

Thanks

**Comment by elieobeid7** [comment]
res.redirect is an express thing In any case, if you want to redirect without express

https://stackoverflow.com/questions/17898183/node-js-301-redirect-non-www-without-express

Check the second answer.

**Comment by dalisoft** [comment]
Solved. Answer [here](https://stackoverflow.com/a/25236904)
Thanks @elieobeid7

**Comment by dalisoft** [comment]
Reopen because didn't solved as getting absolute `url` isn't possible

**Comment by elieobeid7** [comment]
You're trying to redirect the frontend right? just do

`  window.location.href = "https://example.com/" `

**Comment by dalisoft** [comment]
@elieobeid7 No, only backend. I know how-to redirect front

**Comment by elieobeid7** [comment]
You could use this

https://www.npmjs.com/package/redirekt

**Comment by dalisoft** [comment]
I can redirect to any url, except for backend, i want get url of `my backend` or having similar to `.redirect` which redirects auto `/path` to `http://backend.tld/path`

**Comment by elieobeid7** [comment]
Why don't you use express if you think express would solve the issue? It's okay to just use it for redirect, or maybe ask on stackoverflow.

I'm not sure the issue here but I'm not sure, I'm new here. What I would do is install express if I'm sure that express solves the problem.

You see this library is an alternative to socket.io, in socket.io you can't do redirects, in websockets you can't do that either, so I don't think this library can do that, but the owner knows more than I do, so you could wait for his reply about what this library can and cannot do.

**Comment by dalisoft** [comment]
This library does more than **Socket.io** does and @alexhultman made the fastest web-server for Node.js ever have (fix me if i'm wrong).
I'm built new Express alternative, you can see [here](https://github.com/dalisoft/nanoexpress) and it's fast, not fast as uWebSockets.js, but very fast.
I'm trying give uWebSockets.js users familiar API (Express-like API) and then users can **move** to uWebSockets easier, because they see source code and understands how it works, how it wrapped and etc.

**Comment by elieobeid7** [comment]
Ah kudos! hopefully alex would help you more than I can :)

**Comment by dalisoft** [comment]
I hope too, thank you for help @elieobeid7 :)

**Comment by elieobeid7** [comment]
btw in the example i showed 

https://stackoverflow.com/questions/17898183/node-js-301-redirect-non-www-without-express/25236904#25236904

the guy redirects to an absolute url, not sure why it's isn't working with you

`    res.writeHead(301, {"Location": "http://example.com"});`

**Comment by dalisoft** [comment]
The problem comes when i'm trying to use `relative urls`, because uWebSockets.js `req.getUrl()` doesn't return the `Absolute URL`, an `Relative URL`.
For `.redirect` need `Absolute URL` and where i can get the `Absolute URL`?

**Comment by elieobeid7** [comment]
So you want something like ` req.get('host')` in express so you could do `req.get('host') + req.getUrl()`

so ` req.get('host')` not available?


Alright until alex comes and fix it, here's a little hack, we know that the library runs on localhost + port

What we could do is this

```
const port = 3000;
const host = 'http://localhost:' +  port.toString() + '/'  // Im not sure if the last backslash is needed or not
const absolute_url = host + req.getUrl()
```
basically if `req.getUrl()` gives you something like this

`/bla/`

then you don't need /, but if it gives you `bla/` then you need that.

**Comment by dalisoft** [comment]
Thanks, little hack already i am did until i am found fix or author answers

**Comment by ghost** [comment]
From what I understand a redirect is just a response with that 300-ish status and a header with the URL. In that case you already have all the tools for this.

**Comment by dalisoft** [comment]
Thanks. If wontfix, it is not bad, not everyone uses this method.

**Comment by Xetera** [comment]
In case anyone is struggling with redirects, the order of the calls matter.
```js
res
  .writeHeader("Location", "https://example.com")
  .writeStatus("302")
  .end();
```
Will not work but 
```js
res
  .writeStatus("302")
  .writeHeader("Location", "https://example.com")
  .end();
```
will. Not sure if that's intentional but it seems to be the case.

**Comment by ghost** [comment]
First writeHeader calls writeStatus if not already done, so order matters yes

**Comment by ghost** [comment, marked as answer]
I think this thread had been answered - it takes a few lines of string concatenation but you do certainly have all the tools. It could be added as a built-in feature if more people would report this but I think it is kind of a high level thing.

**Reply by dalisoft** [reply]
uWebSockets.js built-in methods are really enough to build full-featured backends and this is not much necessary feature, who needs, some of frameworks which built on top of uWebSockets.js already polyfilled and can be found on GitHub
